domain/src/test/java/org/oppia/domain/oppialogger/exceptions/ExceptionsControllerTest.kt

```diff
@@ -0,0 +1,350 @@
+import org.oppia.app.model.OppiaExceptionLogs
+import org.oppia.domain.oppialogger.ExceptionLogStorageCacheSize
+import org.oppia.testing.FakeExceptionLogger
+import org.oppia.testing.TestCoroutineDispatchers
+import org.oppia.testing.TestDispatcherModule
+import org.oppia.testing.TestLogReportingModule
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.logging.EnableConsoleLog
+import org.oppia.util.logging.EnableFileLog
+import org.oppia.util.logging.GlobalLogLevel
+import org.oppia.util.logging.LogLevel
+import org.oppia.util.networking.NetworkConnectionUtil
+import org.robolectric.annotation.Config
+import javax.inject.Inject
+import javax.inject.Qualifier
+import javax.inject.Singleton
+
+const val TEST_TIMESTAMP = 1556094120000
+
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class ExceptionsControllerTest {
+
+  @Rule
+  @JvmField
+  val mockitoRule: MockitoRule = MockitoJUnit.rule()
+
+  @Inject
+  lateinit var exceptionsController: ExceptionsController
+
+  @Inject
+  lateinit var networkConnectionUtil: NetworkConnectionUtil
+
+  @Inject
+  lateinit var fakeExceptionLogger: FakeExceptionLogger
+
+  @InternalCoroutinesApi
+  @Inject
+  lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
+
+  @Mock
+  lateinit var mockOppiaExceptionLogsObserver: Observer<AsyncResult<OppiaExceptionLogs>>
+
+  @Captor
+  lateinit var oppiaExceptionLogsResultCaptor: ArgumentCaptor<AsyncResult<OppiaExceptionLogs>>
+
+  @Before
+  fun setUp() {
+    networkConnectionUtil = NetworkConnectionUtil(ApplicationProvider.getApplicationContext())
+    setUpTestApplicationComponent()
+  }
+
+  @Test
+  fun testController_logException_NonFatal_logsToRemoteService() {
+    val exceptionThrown = Exception("TEST MESSAGE", Throwable())
+    exceptionsController.logNonFatalException(exceptionThrown, TEST_TIMESTAMP)
+
+    val exceptionLogged = fakeExceptionLogger.getMostRecentException()
+
+    assertThat(exceptionLogged).isEqualTo(exceptionThrown)
+  }
+
+  @Test
+  fun testController_logException_Fatal_logsToRemoteService() {
+    val exceptionThrown = Exception("TEST MESSAGE", Throwable())
+    exceptionsController.logFatalException(exceptionThrown, TEST_TIMESTAMP)
+
+    val exceptionLogged = fakeExceptionLogger.getMostRecentException()
+
+    assertThat(exceptionLogged).isEqualTo(exceptionThrown)
+  }
+
+  // TODO(#1106): Addition of tests tracking behaviour of the controller after uploading of logs to the remote service.
+
+  @ExperimentalCoroutinesApi
+  @InternalCoroutinesApi
+  @Test
+  fun testController_logException_NonFatal_withNoNetwork_logsToCacheStore() {
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.NONE)
+    val exceptionThrown = Exception("TEST MESSAGE", Throwable("TEST CAUSE"))
+    exceptionsController.logNonFatalException(exceptionThrown, TEST_TIMESTAMP)
+
+    val cachedExceptions = exceptionsController.getExceptionLogs()
```

Comment:
Out of curiosity, will ``getExceptionLogs`` be used by anything other than tests?

Reply:
Yes, when we implement the uploading functionality.

Reply:
So my basic motive of using this method was to get the logs without going into threading. Data Providers does that but it's usage around the codebase for retrieval of cache was typically using LiveData only. You're right that the work manager (uploading functionality) won't be using any of the LiveData things. It would simply like to read from the store and upload.

Reply:
Everyone that we do that requires background or async processing goes through DataProviders. We convert DataProviders to LiveData only for when we want to observe the DataProvider from a UI. If this needs to be a DataProvider, instead, you can still test adding listeners to it (see DataProvidersTest fro an example).

Reply:
PTAL